pennylane/templates/subroutines/qsvt.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the QSVT template and qsvt wrapper function.
+"""
+# pylint: disable=too-many-arguments
+import copy
+import pennylane as qml
+from pennylane.queuing import QueuingManager
+from pennylane.ops import BlockEncode, PCPhase
+from pennylane.ops.op_math import adjoint
+from pennylane.operation import AnyWires, Operation
+from pennylane import numpy as np
+
+
+def qsvt(A, angles, wires, convention=None):
+    r"""Implements the
+    `quantum singular value transformation <https://arxiv.org/abs/1806.01838>`__ circuit using
+    :class:`~.BlockEncode` and :class:`~.PCPhase`.
+
+    Given a matrix :math:`A`, and a list of angles :math:`\vec{\phi}`, this template applies a
+    circuit for the quantum singular value transformation as follows.
+
+    When the number of angles is even (:math:`d` is odd), the QSVT circuit is defined as:
+
+    .. math::
+
+        U_{QSVT} = \Pi_{\phi_1}U\left[\prod^{(d-1)/2}_{k=1}\Pi_{\phi_{2k}}U^\dagger
+        \Pi_{\phi_{2k+1}}U\right]\Pi_{\phi_{d+1}},
+
+
+    and when the number of angles is odd (:math:`d` is even):
+
+    .. math::
+
+        U_{QSVT} = \left[\prod^{d/2}_{k=1}\Pi_{\phi_{2k-1}}U^\dagger\Pi_{\phi_{2k}}U\right]
+        \Pi_{\phi_{d+1}}.
+
+    Here, :math:`U` denotes a block encoding of :math:`A` via :class:`~.BlockEncode` and
+    :math:`\Pi_\phi` denotes a projector-controlled phase shift with angle :math:`\phi`
+    via :class:`~.PCPhase`.
+
+    This circuit applies a polynomial transformation (:math:`Poly^{SV}`) to the singular values of
+    the block encoded matrix:
+
+    .. math::
+
+        \begin{align}
+             U_{QSVT}(A, \phi) &=
+             \begin{bmatrix}
+                Poly^{SV}(A) & \cdot \\
+                \cdot & \cdot
+            \end{bmatrix}.
+        \end{align}
+
+    .. seealso::
+
+        :class:`~.QSVT` and `A Grand Unification of Quantum Algorithms <https://arxiv.org/pdf/2105.02859.pdf>`_
+
+    Args:
+        A (tensor_like): the general :math:`(n \times m)` matrix to be encoded
+        angles (tensor_like): a list of angles by which to shift to obtain the desired polynomial
+        wires (Iterable): the wires the template acts on
+        convention (string): can be set to "Wx" to convert quantum signal processing angles in the
```

Comment:
I think we should explain somewhere what the different conventions are. Maybe a "theory" section below? Not sure.